File: android/src/main/java/com/guhungry/expophotomanipulator/ExpoPhotoManipulatorModule.kt
package com.guhungry.expophotomanipulator

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Paint
import android.graphics.Typeface
import com.facebook.react.common.assets.ReactFontManager
import com.guhungry.expophotomanipulator.utils.ImageUtils.bitmapFromUri
import com.guhungry.expophotomanipulator.utils.ImageUtils.cropBitmapFromUri
import com.guhungry.expophotomanipulator.utils.ImageUtils.mutableOptions
import com.guhungry.expophotomanipulator.utils.ImageUtils.saveTempFile
import com.guhungry.expophotomanipulator.utils.ParamUtils
import com.guhungry.expophotomanipulator.utils.ParamUtils.flipMode
import com.guhungry.expophotomanipulator.utils.ParamUtils.pointF
import com.guhungry.expophotomanipulator.utils.ParamUtils.rotationMode
import com.guhungry.expophotomanipulator.utils.ParamUtils.textOptions
import com.guhungry.photomanipulator.BitmapUtils.flip
import com.guhungry.photomanipulator.BitmapUtils.overlay
import com.guhungry.photomanipulator.BitmapUtils.printText
import com.guhungry.photomanipulator.BitmapUtils.rotate
import com.guhungry.photomanipulator.MimeUtils
import com.guhungry.photomanipulator.model.TextStyle
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition


class ExpoPhotoManipulatorModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("ExpoPhotoManipulator")

    Constants()

    AsyncFunction("batch") { uri: String, operations: List<Map<String, Any>>, cropRegion: Rect, targetSize: Size?, quality: Double?, mimeType: String?, promise: Promise ->
      try {
        var output = cropBitmapFromUri(context, uri, cropRegion.toCGRect(), targetSize?.toCGSize())

        // Operations
        for (operation in operations) {
          output = processBatchOperation(output, operation)
        }

        // Save & Optimize
        val file = saveTempFile(context, output, mimeType!!, FILE_PREFIX, quality!!.toInt())
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("crop") { uri: String, cropRegion: Rect, targetSize: Size?, mimeType: String?, promise: Promise ->
      try {
        val output = cropBitmapFromUri(context, uri, cropRegion.toCGRect(), targetSize?.toCGSize())

        val file = saveTempFile(context, output, mimeType!!, FILE_PREFIX, DEFAULT_QUALITY)
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("flipImage") { uri: String, mode: String, mimeType: String?, promise: Promise ->
      try {
        val input = bitmapFromUri(context, uri, mutableOptions())

        val output = flip(input, flipMode(mode))
        input.recycle()

        val file = saveTempFile(context, output, mimeType!!, FILE_PREFIX, DEFAULT_QUALITY)
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("rotateImage") { uri: String, mode: String, mimeType: String?, promise: Promise ->
      try {
        val input = bitmapFromUri(context, uri, mutableOptions())

        val output = rotate(input, rotationMode(mode))
        input.recycle()

        val file = saveTempFile(context, output, mimeType!!, FILE_PREFIX, DEFAULT_QUALITY)
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("overlayImage") { uri: String, icon: String, position: Point, mimeType: String?, promise: Promise ->
      try {
        val output = bitmapFromUri(context, uri, mutableOptions())
        val overlay = bitmapFromUri(context, icon)

        overlay(output, overlay, position.toPointF())
        overlay.recycle()

        val file = saveTempFile(context, output, mimeType!!, FILE_PREFIX, DEFAULT_QUALITY)
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("printText") { uri: String, texts: List<TextOptions>, mimeType: String?, promise: Promise ->
      try {
        val output = bitmapFromUri(context, uri, mutableOptions())

        for (text in texts) {
          printLine(output, text)
        }

        val file = saveTempFile(context, output, mimeType!!, FILE_PREFIX, DEFAULT_QUALITY)
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }

    AsyncFunction("optimize") { uri: String, quality: Double, promise: Promise ->
      try {
        val output = bitmapFromUri(context, uri)

        val file = saveTempFile(context, output, MimeUtils.JPEG, FILE_PREFIX, quality.toInt())
        output.recycle()

        promise.resolve(file)
      } catch (e: Exception) {
        promise.reject(CodedException(e))
      }
    }
  }

  private fun processBatchOperation(image: Bitmap, operation: Map<String, Any>?): Bitmap {
    if (operation == null) return image
    when (operation["operation"] as? String) {
      "text" -> {
        val text = operation["options"] as? Map<String, Any> ?: return image

        printLine(image, textOptions(text))
        return image
      }

      "overlay" -> {
        val uri = operation["overlay"] as? String ?: return image

        val overlay = bitmapFromUri(context, uri)
        overlay(image, overlay, pointF(operation["position"] as? Map<String, Any>)!!)
        return image
      }

      "flip" -> {
        return flip(image, flipMode(operation["mode"] as String))
      }

      "rotate" -> {
        return rotate(image, rotationMode(operation["mode"] as String))
      }

      else -> return image
    }
  }

  private fun printLine(image: Bitmap, options: TextOptions) {
    val style = TextStyle(
      options.color.toColor(),
      options.textSize,
      getFont(options.fontName),
      Paint.Align.LEFT,
      options.thickness,
      options.rotation,
      options.shadowRadius,
      options.shadowOffset?.x ?: 0f,
      options.shadowOffset?.y ?: 0f,
      options.shadowColor?.toColor()
    )
    printText(image, options.text, options.position.toPointF(), style)
  }

  private fun getFont(fontName: String?): Typeface {
    if (fontName == null) return Typeface.DEFAULT
    return try {
      ReactFontManager.getInstance()
        .getTypeface(fontName, Typeface.NORMAL, context.assets)
    } catch (e: Exception) {
      Typeface.DEFAULT
    }
  }

  companion object {
    private const val FILE_PREFIX = "EPM_"
    private const val DEFAULT_QUALITY = 100
  }
}

File: ios/ExpoPhotoManipulatorModule.swift
import ExpoModulesCore
import WCPhotoManipulator

public class ExpoPhotoManipulatorModule: Module {
    typealias LoadImageResult = (Result<UIImage, Error>) -> Void
    internal static let defaultQuality = CGFloat(100)
    
    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    public func definition() -> ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('ExpoPhotoManipulator')` in JavaScript.
        Name("ExpoPhotoManipulator")

        // Sets constant properties on the module. Can take a dictionary or a closure that returns a dictionary.
        Constants([:])

        AsyncFunction("batch", batch)
        AsyncFunction("crop", crop)
        AsyncFunction("flipImage", flipImage)
        AsyncFunction("rotateImage", rotateImage)
        AsyncFunction("overlayImage", overlayImage)
        AsyncFunction("printText", printText)
        AsyncFunction("optimize", optimize)
    }
    
    internal func batch(_ url: URL, operations: [[String:Any]], cropRegion: Rect, targetSize: Size?, quality: CGFloat, mimeType: String, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(let image):
                guard var source = self.cropOrResize(image, cropRegion, targetSize) else {
                    promise.resolve(ImageOperationException())
                    return
                }
                
                for operation in operations {
                    guard let output = self.performOperation(source, operation) else {
                        promise.resolve(ImageOperationException())
                        return
                    }
                    source = output
                }
                
                self.saveFile(source, mimeType: mimeType, quality: ExpoPhotoManipulatorModule.defaultQuality, promise: promise)
            }
        }
    }
    
    private func performOperation(_ image: UIImage, _ operation: [String:Any]) -> UIImage? {
        guard let type = operation["operation"] as? String else {
            return image
        }
        switch type {
        case "text":
            guard let text = ParamUtils.textOptions(operation["options"]) else {
                return image
            }
            return image.drawText(text.text, position: text.getPosition(), style: text.getTextStyle())
        case "flip":
                return image.flip(ParamUtils.flipMode(operation["mode"]))
        case "rotate":
            return image.rotate(ParamUtils.rotationMode(operation["mode"]))
        case "overlay":
            guard let url = ParamUtils.url(operation["overlay"]) else {
                return image
            }
            guard let overlay = ImageUtils.imageFromUrl(url) else {
                return image
            }
            guard let position = ParamUtils.point(operation["position"]) else {
                return image
            }
            
            return image.overlayImage(overlay, position: position.toPoint())
        default:
            return image
        }
    }
    
    internal func crop(_ url: URL, cropRegion: Rect, targetSize: Size?, mimeType: String, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(let image):
                guard let result = self.cropOrResize(image, cropRegion, targetSize) else {
                    promise.resolve(ImageOperationException())
                    return
                }

                self.saveFile(result, mimeType: mimeType, quality: ExpoPhotoManipulatorModule.defaultQuality, promise: promise)
            }
        }
    }
    
    private func cropOrResize(_ image: UIImage, _ cropRegion: Rect, _ targetSize: Size?) -> UIImage? {
        guard let targetSize = targetSize else {
            return image.crop(cropRegion.toCGRect())
        }
        return image.crop(cropRegion.toCGRect(), targetSize: targetSize.toCGSize())
    }
    
    internal func flipImage(_ url: URL, flipMode: String, mimeType: String, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(let image):
                let result = image.flip(ParamUtils.flipMode(flipMode))

                self.saveFile(result, mimeType: mimeType, quality: ExpoPhotoManipulatorModule.defaultQuality, promise: promise)
            }
        }
    }
    
    internal func rotateImage(_ url: URL, rotationMode: String, mimeType: String, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(let image):
                let result = image.rotate(ParamUtils.rotationMode(rotationMode))

                self.saveFile(result, mimeType: mimeType, quality: ExpoPhotoManipulatorModule.defaultQuality, promise: promise)
            }
        }
    }
    
    internal func overlayImage(_ url: URL, overlay: URL, position: Point, mimeType: String, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(let image):
                self.loadImage(overlay) { result in
                    switch result {
                    case .failure(let error):
                        promise.reject(error)
                    case .success(let icon):
                        guard let output = image.overlayImage(icon, position: position.toPoint()) else {
                            promise.resolve(ImageOperationException())
                            return
                        }

                        self.saveFile(output, mimeType: mimeType, quality: ExpoPhotoManipulatorModule.defaultQuality, promise: promise)
                    }
                }
            }
        }
    }
    
    internal func printText(_ url: URL, texts: [TextOptions], mimeType: String, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(var image):
                for option in texts {
                    let text = option.text
                    let position = option.getPosition()
                    let style = option.getTextStyle()

                    guard let result = image.drawText(text, position: position, style: style) else {
                        promise.resolve(ImageOperationException())
                        return
                    }
                    image = result
                }

                self.saveFile(image, mimeType: mimeType, quality: ExpoPhotoManipulatorModule.defaultQuality, promise: promise)
            }
        }
    }

    internal func optimize(_ url: URL, quality: CGFloat, promise: Promise) {
        self.loadImage(url) { result in
            switch result {
            case .failure(let error):
                promise.reject(error)
            case .success(let image):
                self.saveFile(image, mimeType: MimeUtils.JPEG, quality: quality, promise: promise)
            }
        }
    }
    
    internal func loadImage(_ url: URL, callback: @escaping LoadImageResult) {
        guard let image = ImageUtils.imageFromUrl(url) else {
            callback(.failure(ImageNotFoundException()))
            return
        }
        callback(.success(image))
    }
    
    internal func saveFile(_ image: UIImage, mimeType: String, quality: CGFloat, promise: Promise) {
        guard let output = ImageUtils.saveTempFile(image, mimeType: mimeType, quality: quality) else {
            promise.reject(SaveFileException())
            return
        }
    
        promise.resolve(output)
    }
}

